examples/TestAppUwp/Model/SessionModel.cs

```diff
@@ -0,0 +1,998 @@
+                IceDataSeparator = "|"
+            };
+            NodeDssSignaler.SendMessageAsync(message);
+        }
+
+        private void OnIceStateChanged(IceConnectionState newState)
+        {
+            Logger.Log($"Connection state changed to {newState}.");
+            IceConnectionState = newState;
+
+            // ICE was disconnected, which generally indicates that the remote peer was closed.
+            // Shut down the remote media player and clear the remote video statistics.
+            if (newState == IceConnectionState.Disconnected)
+            {
+                //OnPeerDisconnected();
+            }
+        }
+
+        private void OnIceGatheringStateChanged(IceGatheringState newState)
+        {
+            Logger.Log($"ICE gathering changed to {newState}.");
+            IceGatheringState = newState;
+        }
+
+        private void OnPeerDisconnected()
+        {
+            Logger.Log($"Peer disconnected.");
+
+            //// Update the video source if needed
+            //var selectedVideoTrack = LocalVideoTracks.SelectedItem;
+            //if (selectedVideoTrack.IsRemote)
+            //{
+
+            //    _remoteVideoPlayer.Pause();
+            //    _remoteVideoPlayer.Source = null;
+            //    remoteVideo.SetMediaPlayer(null);
+            //    _remoteVideoSource?.NotifyError(MediaStreamSourceErrorStatus.ConnectionToServerLost);
+            //    _remoteMediaSource?.Dispose();
+            //}
+
+            // Remove all remote tracks
+            for (int i = AudioTracks.Count - 1; i >= 0; --i)
+            {
+                if (AudioTracks[i].IsRemote)
+                {
+                    AudioTracks.RemoveAt(i);
+                }
+            }
+            for (int i = VideoTracks.Count - 1; i >= 0; --i)
+            {
+                if (VideoTracks[i].IsRemote)
+                {
+                    VideoTracks.RemoveAt(i);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Notify the peer connection that a local SDP offer message was applied.
+        /// </summary>
+        public void NotifyLocalOfferApplied()
+        {
+            lock (_stateLock)
+            {
+                Debug.Assert(_negotiationState == NegotiationState.Starting);
+                _negotiationState = NegotiationState.HaveLocalOffer;
+            }
+            RaisePropertyChanged("NegotiationState");
+        }
+
+        /// <summary>
+        /// Notify the peer connection that a local SDP answer message was applied.
+        /// </summary>
+        public void NotifyLocalAnswerApplied()
+        {
+            lock (_stateLock)
+            {
+                Debug.Assert(_negotiationState == NegotiationState.HaveRemoteOffer);
+                _negotiationState = NegotiationState.Stable;
+            }
+            RaisePropertyChanged("NegotiationState");
+            RaisePropertyChanged("CanNegotiate");
```

Comment:
nit: add a `private set` to `NegotiationState` that raises these events instead of duplicating these lines everywhere

Reply:
Hmm, it might not be easy to always deal with the lock correctly though, so feel free to ignore - maybe just add a comment to `_negotiationState` reminding maintainers to raise the events on change.

Reply:
Fixed. I added a single private method to avoid duplicating the code, and explicitly coded the reason why `CanNegotiate` is only raised sometimes, which was not obvious in the previous code.